pkg/controller/elasticsearch/remotecluster/remoteca/watches.go

```diff
@@ -0,0 +1,108 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package remoteca
+
+import (
+	"fmt"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
+
+	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/watches"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/certificates/transport"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// addWatches set watches on objects needed to manage the association between a local and a remote cluster.
+func addWatches(c controller.Controller, r *ReconcileRemoteCa) error {
+	// Watch for changes to RemoteCluster
+	if err := c.Watch(&source.Kind{Type: &esv1.Elasticsearch{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	// Watch Secrets that contain remote certificate authorities managed by this controller
+	if err := c.Watch(&source.Kind{Type: &v1.Secret{}}, &handler.EnqueueRequestsFromMapFunc{
+		ToRequests: newToRequestsFuncFromSecret(),
+	}); err != nil {
+		return err
+	}
+
+	// Dynamically watches the certificate authorities involved in a cluster relationship
+	if err := c.Watch(&source.Kind{Type: &v1.Secret{}}, r.watches.Secrets); err != nil {
+		return err
+	}
+	if err := r.watches.Secrets.AddHandlers(
+		&watches.OwnerWatch{
+			EnqueueRequestForOwner: handler.EnqueueRequestForOwner{
+				IsController: true,
+				OwnerType:    &esv1.Elasticsearch{},
+			},
+		},
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newToRequestsFuncFromSecret creates a watch handler function that creates reconcile requests based on the
+// labels set on a Secret which contains the remote CA.
+func newToRequestsFuncFromSecret() handler.ToRequestsFunc {
+	return func(obj handler.MapObject) []reconcile.Request {
+		labels := obj.Meta.GetLabels()
+		if secretType, ok := labels[common.TypeLabelName]; !ok || secretType != TypeLabelValue {
+			return nil
+		}
+		clusterAssociationName, ok := labels[RemoteClusterNameLabelName]
+		if !ok {
+			return nil
+		}
+		clusterAssociationNamespace, ok := labels[RemoteClusterNamespaceLabelName]
+		if !ok {
+			return nil
+		}
+		return []reconcile.Request{
+			{NamespacedName: types.NamespacedName{
+				Namespace: clusterAssociationNamespace,
+				Name:      clusterAssociationName},
+			},
+		}
+	}
+}
+
+func watchName(local types.NamespacedName, remote types.NamespacedName) string {
+	return fmt.Sprintf(
+		"%s-%s-%s-%s",
+		local.Namespace,
+		local.Name,
+		remote.Namespace,
+		remote.Name,
+	)
+}
+
+// addCertificatesAuthorityWatches sets some watches on all secrets containing the certificate of a CA involved in a association.
```

Comment:
Because I had to think about it twice (local vs. remote, what is watched?), maybe worth a comment like:

```
// watch the remote cluster CA secret to reconcile the local cluster CA secrets
```

Reply:
Both are watched:
* The local CA (from the ES of the Spec) is watched to update the trusted certificates in the remote clusters.
* The remote CAs (from the ES clusters referenced in `Spec.RemoteCluster`) are watched to update the trusted certificates of the local cluster.